### PR TITLE
automate-gateway: clean up iam v1/v2 switching

### PR DESCRIPTION
**Pull request by srenatus**
### :nut_and_bolt: Description: What code changed, and why?

There's been remnants of the automatic IAM v1/v2 switching in automate-gateway that are no longer needed now that v1 is gone.

Thus, this removes the extra layers:

- `pkg/authorizer` is gone
- `SwitchingAuthorizationHandler` has been removed, all calls now go _directly_ to the v2 handler code

### :chains: Related Resources

- #3272 

### :+1: Definition of Done

Still works. Less code.

### :athletic_shoe: How to Build and Test the Change

- deployinate, `rebuild components/automate-gateway`
- click around in the UI -- who are we kidding? see tests pass, be confident that we've exercised enough from it.

### :white_check_mark: Checklist

- [X] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [X] Tests added/updated?
- [ ] Docs added/updated?
- [X] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### :camera: Screenshots, if applicable

NA

### CODE REVIEW

File: components/automate-gateway/gateway/helpers_internal_test.go [hunk 1]
```diff
@@ -116,7 +115,7 @@ func newMockGatewayServer(t *testing.T, services ...interface{}) Server {
 
 	gw := New(cfg)
 	gw.clientsFactory = mockClientsFactory
-	gw.authorizer = authorizer.NewAuthorizer(auth.AuthorizationHandler(mockAuthorizationClient))
+	gw.authorizer = auth.AuthorizationHandler(mockAuthorizationClient)
```
Comment (srenatus) [inline components/automate-gateway/gateway/helpers_internal_test.go:118]: here, the extra layer is removed -- we directly call the authv2 implementation

File: components/automate-gateway/gateway/middleware/auth_interceptor.go [hunk 2]
```diff
@@ -37,60 +36,6 @@ func NewAuthInterceptor(
 	return &authInterceptor{authn: authn, authz: authz}
 }
 
-type SwitchingAuthorizationHandler interface {
```
Comment (srenatus) [inline components/automate-gateway/gateway/middleware/auth_interceptor.go:40]: The unnecessary types are removed, the rest is moved to the already-present `types.go` in the same directory.

File: components/automate-gateway/gateway/middleware/authv2/authv2.go [hunk 3]
```diff
@@ -82,21 +80,9 @@ func (c *client) Handle(ctx context.Context, subjects []string, projectsToFilter
...
-}
-
-func (r *resp) GetAuthorized() bool {
-	return r.authorized
-}
-
+// IsAuthorized takes care of HTTP authorization in the custom HTTP handlers
 func (c *client) IsAuthorized(ctx context.Context, subjects []string, resource, action string,
-	projectsToFilter []string) (middleware.AuthorizationResponse, error) {
+	projectsToFilter []string) (context.Context, bool, error) {
```
Comment (srenatus) [inline components/automate-gateway/gateway/middleware/authv2/authv2.go:86]: Inlined the response, a simplification available now that we don't have to unify two different PB responses that both have a `GetAuthorized` method.

File: components/automate-gateway/handler/iam/v2/introspect/introspect.go [hunk 4]
```diff
@@ -232,13 +232,15 @@ func (a *AuthzServer) getAllowedMap(
 	// Fetches the id of the current user PLUS the team ids for that user
 	subjects := auth_context.FromContext(ctx).Subjects
 
-	resp, err := a.filterHandler.FilterAuthorizedPairs(ctx, subjects, mapByResourceAndAction, methodsInfo)
+	inputPairs := pairs.GetKeys(mapByResourceAndAction)
+
+	filteredPairs, err := a.introspectionHandler.FilterAuthorizedPairs(ctx, subjects, inputPairs)
```
Comment (srenatus) [inline components/automate-gateway/handler/iam/v2/introspect/introspect.go:237]: There's only one set of `mapByResourceAndAction` and `methodsInfo`, so they're no longer passed in, but retrieved in the introspection code.

File: components/automate-gateway/gateway/middleware/auth_interceptor.go [hunk 5]
```diff
@@ -37,60 +36,6 @@ func NewAuthInterceptor(
...
-	) (*FilterPairsResponse, error)
-}
-
-type FilterPairsResponse struct {
-	Pairs                  []*pairs.Pair
-	MethodsInfo            map[string]pairs.Info
-	MapByResourceAndAction map[pairs.Pair][]string
-}
-
-type FilterProjectsResponse struct {
```
Comment (srenatus) [inline components/automate-gateway/gateway/middleware/auth_interceptor.go:60]: ❓ This type was *unused* -- and there's not much project-iness in the introspection code either. I haven't fully understood the situation... is this in order? @msorens, would you know more about this? What am I missing?

Comment (srenatus) [reply components/automate-gateway/gateway/middleware/auth_interceptor.go:60]: https://buildkite.com/chef/chef-automate-master-verify-private/builds/11412#d96352e3-6c8b-4a5f-84fe-3b32cdbfe439/441-1271 👈 So cypress fails on some project-related stuff. I will revisit this, I suppose. At any rate, if someone more familiar with this knows where I took the wrong turn; or if these failures really are "just flakiness", I'd appreciate the insights.

Comment (tylercloke) [reply components/automate-gateway/gateway/middleware/auth_interceptor.go:60]: The flaky build of cypress always fails. We have flaky and non-flaky sets of the cypress tests. The delta in flaky is either known to sometimes fail or is broken and a WIP. Nothing in there for your build looks out of the ordinary to me.

File: components/automate-gateway/gateway/middleware/authv2/authv2.go [hunk 6]
```diff
@@ -106,21 +93,21 @@ func (c *client) IsAuthorized(ctx context.Context, subjects []string, resource,
...
+		return nil, false, errors.Errorf("error authorizing action %q on resource %q for members %q: %s",
 			action, resource, subjects, err.Error())
 	}
 	projects := filteredResp.Projects
 
-	return &resp{
-		ctx:        auth_context.NewContext(ctx, subjects, projects, resource, action),
-		authorized: len(projects) != 0,
-	}, nil
+	return auth_context.NewContext(ctx, subjects, projects, resource, action),
```
Comment (msorens) [inline components/automate-gateway/gateway/middleware/authv2/authv2.go:105]: Would appreciate a comment here (or just above the function signature above) explaining the return values, most importantly that the boolean being returned indicates authorized or not, and that equates to the presence of one or more projects.

File: components/automate-gateway/gateway/middleware/authv2/authv2.go (hunk 3 above)
Comment (blakestier) [reply components/automate-gateway/gateway/middleware/authv2/authv2.go:86]: I would find it helpful to have these returned values be named (`isAuthorized bool`)

File: components/automate-gateway/gateway/services.go [hunk 7]
```diff
@@ -874,12 +874,15 @@ func (s *Server) authRequest(r *http.Request, resource, action string) (context.
...
 	}
-
-	// err is nil if authorization succeeded
-	// Note: if we need all the auth info, use auth_context.NewOutgoingContext
-	return auth_context.NewOutgoingProjectsContext(authzResp.Ctx()), authzResp.Err()
+	if authorized {
+		// Note: if we need all the auth info, use auth_context.NewOutgoingContext
+		return auth_context.NewOutgoingProjectsContext(newCtx), nil
+	}
+	return nil, errors.Errorf("subjects %v are not permitted to do action %v to resource %v in projects %v",
```
Comment (msorens) [inline components/automate-gateway/gateway/services.go:886]: Compare the text to components/automate-gateway/gateway/middleware/authv2/authv2.go, L71 and L76. Should we use the same verbiage? Not sure. Different verbiage allows an unambiguous grep in the log if needed. However, might at least one to be consistent on `%v` vs `%q`. Does `%v` deliver less escaped quotes...?

File: components/automate-gateway/gateway/services.go [hunk 8]
```diff
@@ -874,12 +874,15 @@ func (s *Server) authRequest(r *http.Request, resource, action string) (context.
 
 	projects := auth_context.ProjectsFromMetadata(md)
 
-	authzResp, err := s.authorizer.IsAuthorized(ctx, subjects, resource, action, projects)
+	newCtx, authorized, err := s.authorizer.IsAuthorized(ctx, subjects, resource, action, projects)
 	if err != nil {
+		// any error is handled as "permission denied" at the call sites, with the error text returned
```
Comment (blakestier) [inline components/automate-gateway/gateway/services.go:879]: When I read this, I think "ok, so if there's an error, it's unauthorized, but also, it's unauthorized if the bool `authorized ` is false . . . 

Are we really saying "an err is returned if authorization can't be determined. Upstream services, however, will consider it equivalent to an explicit permission denied, with error message included"? (words up for refinement, just trying to get the meaning)

File: components/automate-gateway/gateway/services.go [hunk 9]
```diff
@@ -874,12 +874,15 @@ func (s *Server) authRequest(r *http.Request, resource, action string) (context.
...
 		return nil, errors.Wrap(err, "authz-service error")
 	}
-
-	// err is nil if authorization succeeded
-	// Note: if we need all the auth info, use auth_context.NewOutgoingContext
-	return auth_context.NewOutgoingProjectsContext(authzResp.Ctx()), authzResp.Err()
+	if authorized {
+		// Note: if we need all the auth info, use auth_context.NewOutgoingContext
```
Comment (blakestier) [inline components/automate-gateway/gateway/services.go:883]: this is like, a "someday if we want" note?

Comment (blakestier) [reply components/automate-gateway/gateway/services.go:883]: ( i know you didn't write it, just wonderin )

File: components/automate-gateway/handler/iam/v2/introspect/introspect.go [hunk 10]
```diff
@@ -29,18 +29,18 @@ import (
...
-	filterHandler middleware.SwitchingFilterHandler
+	client               authz.AuthorizationClient // now only needed for GetVersion
+	introspectionHandler middleware.IntrospectionHandler
 }
 
 // NewServer creates a server with its client
 func NewServer(
 	client authz.AuthorizationClient,
-	filterHandler middleware.SwitchingFilterHandler,
+	IntrospectionHandler middleware.IntrospectionHandler,
```
Comment (blakestier) [inline components/automate-gateway/handler/iam/v2/introspect/introspect.go:39]: Does this need to be capitalized?

File: components/automate-gateway/handler/iam/v2/introspect/introspect_test.go [hunk 11]
```diff
@@ -257,5 +256,5 @@ func testServerAndHandler(t *testing.T) (
 		s,
 		introspect.NewServer(
 			nil,
-			authorizer.NewAuthorizer(authv2.AuthorizationHandler(client)))
+			authv2.AuthorizationHandler(client))
```
Comment (blakestier) [inline components/automate-gateway/handler/iam/v2/introspect/introspect_test.go:259]: we have a card down the road to "neutralize" the v2 stuff (there's no v1, so . . .). I'm not sure it really matters and we could sweep this later, but would be cool to name this import `auth` as a start, so there's no immediate code-reader impulse to differentiate

File: components/automate-gateway/handler/iam/v2/introspect/introspect.go (hunk 10 above)
Comment (srenatus) [reply components/automate-gateway/handler/iam/v2/introspect/introspect.go:39]: Nope, good catch. I am not good at all at search-and-replace. :D

**Comment by srenatus** [comment]
@tylercloke @msorens @blakestier Thanks for the reviews. I'll get your comments addressed tomorrow.

File: components/automate-gateway/gateway/services.go (hunk 9 above)
Comment (srenatus) [reply components/automate-gateway/gateway/services.go:883]: I think it's "if we wanted X, we should do Y here", so, yeah, your interpretation.

File: components/automate-gateway/gateway/middleware/authz/authz.go [hunk 12]
```diff
@@ -106,21 +96,37 @@ func (c *client) IsAuthorized(ctx context.Context, subjects []string, resource,
...
-		authorized: len(projects) != 0,
-	}, nil
+	return auth_context.NewContext(ctx, subjects, projects, resource, action),
+		len(projects) != 0,
+		nil
 }
 
+// FilterAuthorizedPairs drives authz introspection
+// Q: There's no projects here! This can't be right!
+// A(msorens): The short answer to that is introspection is not yet project-aware.
```
Comment (srenatus) [inline components/automate-gateway/gateway/middleware/authz/authz.go:115]: @msorens hope you don't mind that I've liberated your useful comment from slack; it'll be easier to find here for future explorers.

**Comment by srenatus** [comment]
The requested changes seemed fairly uncontroversial; I'll merge when green.

**Pull request merged**